test/config/providers/chat_provider_reaction_test.dart

```diff
@@ -0,0 +1,306 @@
+        messagesToReturn: [testMessage],
+      );
+
+      container = ProviderContainer(
+        overrides: [
+          authProvider.overrideWith(
+            () => MockAuthNotifier(isAuthenticated: true),
+          ),
+          activePubkeyProvider.overrideWith(
+            () => MockActivePubkeyNotifier(testPubkey),
+          ),
+          groupMessagesProvider.overrideWith(
+            () => mockGroupMessagesNotifier,
+          ),
+          chatProvider.overrideWith(
+            () => ChatNotifier(messageSenderService: mockMessageSenderService),
+          ),
+        ],
+      );
+
+      // Initialize the provider and load messages
+      final notifier = container.read(chatProvider.notifier);
+      await notifier.loadMessagesForGroup(testGroupId);
+    });
+
+    tearDown(() {
+      container.dispose();
+    });
+
+    test('updateMessageReaction adds reaction optimistically', () async {
+      final notifier = container.read(chatProvider.notifier);
+      const reactionEmoji = '👍';
+
+      // Verify initial state
+      var messages = container.read(chatProvider).groupMessages[testGroupId];
+      expect(messages!.first.reactions, isEmpty);
+
+      // Add reaction
+      final result = await notifier.updateMessageReaction(
+        message: testMessage,
+        reaction: reactionEmoji,
+      );
+
+      expect(result, true);
+      expect(mockMessageSenderService.reactionCallCount, 1);
+
+      // Verify optimistic update
+      messages = container.read(chatProvider).groupMessages[testGroupId];
+      expect(messages!.first.reactions, isNotEmpty);
+      expect(messages.first.reactions.first.emoji, reactionEmoji);
+      expect(messages.first.reactions.first.user.publicKey, testPubkey);
+    });
+
+    test('updateMessageReaction removes reaction optimistically (toggle)', () async {
+      const reactionEmoji = '👍';
+
+      // Setup: Message already has a reaction from current user
+      final reaction = Reaction(
+        emoji: reactionEmoji,
+        user: User(
+          id: testPubkey,
+          publicKey: testPubkey,
+          displayName: 'Me',
+          nip05: '',
+        ),
+        createdAt: DateTime.now(),
+      );
+
+      final messageWithReaction = testMessage.copyWith(reactions: [reaction]);
+
+      // Update the mock to return this message (simulating state before toggle)
+      // We need to manually update the state because we can't easily inject it into the provider's internal state directly
+      // without going through the load process again or using a method exposed for testing.
+      // However, for this test, we can just call updateMessageReaction on the messageWithReaction.
+      // But wait, the provider looks up the message in its state by ID.
+      // So we need to ensure the provider's state has the message with reaction.
+
+      // Let's re-initialize with the message having a reaction
+      mockGroupMessagesNotifier = MockGroupMessagesNotifier(
```

Comment:
I think you could move this mocks to setUp, inside another group so that you don't mock the providers twice for this test case